### PR TITLE
copr/metrics: record scan efficiency

### PR DESCRIPTION
**Pull request by BusyJay**
@siddontang @disksing @overvenus PTAL

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Should we statistic the keys count in cursor near_seek?

File: src/storage/mvcc/reader.rs [hunk 1]
```diff
@@ -17,6 +17,12 @@ use super::{Error, Result};
 use super::lock::Lock;
 use super::write::{Write, WriteType};
 
+#[derive(Clone, Default, Debug)]
+pub struct ScanMetrics {
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: can we embed this in the MvccReader?

Comment (siddontang) [inline src/storage/mvcc/reader.rs]: @disksing 

How can you use this in scheduler?

File: src/storage/mvcc/reader.rs [hunk 2]
```diff
@@ -17,6 +17,12 @@ use super::{Error, Result};
...
+    pub scanned_keys: u64,
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: do we need to distinguish seek and reverse seek?

File: src/server/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -152,12 +154,25 @@ impl RequestTask {
         let wait_time = self.wait_time.unwrap();
         COPR_REQ_HANDLE_TIME.with_label_values(&["select", type_str])
             .observe(handle_time - wait_time);
+
+        let scanned_keys = self.scan_metrics.scanned_keys as f64;
+        COPR_SCAN_KEYS.with_label_values(&["select", type_str]).observe(scanned_keys);
+        let efficiency = if self.scan_metrics.scanned_keys == 0 {
```
Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: maybe we can use a function in ScanMetrics to calculate efficiency.

Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: scheduler may also use it. /cc @disksing

File: src/storage/mvcc/reader.rs (hunk 1 above)
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: If there is an error during scanning, it's not convenient to get the metrics.

File: src/storage/mvcc/reader.rs (hunk 2 above)
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: All our metrics currently doesn't distinguish these two scenario.

**Comment by BusyJay** [comment]
PTAL

File: src/server/coprocessor/metrics.rs [hunk 4]
```diff
@@ -55,4 +55,20 @@ lazy_static! {
             "Total number of pending push down request.",
             &["type"]
         ).unwrap();
+
+    pub static ref COPR_SCAN_KEYS: HistogramVec =
+        register_histogram_vec!(
+            "tikv_coprocessor_scan_keys",
+            "Bucketed histogram of coprocessor per request scan keys",
+            &["type", "req"],
+            exponential_buckets(0.0005, 2.0, 20).unwrap()
```
Comment (disksing) [inline src/server/coprocessor/metrics.rs]: Make first bucket 1.0 for count.

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/storage/mvcc/reader.rs (hunk 1 above)
Comment (disksing) [inline src/storage/mvcc/reader.rs]: @siddontang Seems only suitable for the Scan command, which is rarely used in kv commands.

**Comment by siddontang** [comment]
Do we need to statistic next count?

**Comment by disksing** [comment]
LGTM

**Comment by BusyJay** [comment]
@disksing @siddontang PTAL

**Comment by disksing** [comment]
LGTM

**Pull request merged**